src/lib.rs

```diff
@@ -578,13 +578,97 @@ fn readline_raw<H: Helper>(
     user_input
 }
 
-fn readline_direct() -> Result<String> {
-    let mut line = String::new();
-    if io::stdin().read_line(&mut line)? > 0 {
-        Ok(line)
-    } else {
-        Err(error::ReadlineError::Eof)
+// Helper to handle backspace characters in a direct input
+fn apply_backspace_direct(input: &str) -> String {
+    // Setup the output buffer
+    // No '\b' in the input in the common case, so set the capacity to the input length
+    let mut out = String::with_capacity(input.len());
+
+    // Keep track of the size of each grapheme from the input
+    // As many graphemes as input bytes in the common case
+    let mut grapheme_sizes: Vec<u8> = Vec::with_capacity(input.len());
+
+    for g in unicode_segmentation::UnicodeSegmentation::graphemes(input, true) {
+        match g {
+            // backspace char
+            "\u{0008}" => {
+                if let Some(n) = grapheme_sizes.pop() {
+                    // Remove the last grapheme
+                    out.truncate(out.len() - n as usize);
+                }
+            }
+            _ => {
+                out.push_str(g);
+                grapheme_sizes.push(g.len() as u8);
+            }
+        }
+    }
+
+    out
+}
+
+fn readline_direct(
+    mut reader: impl BufRead,
+    mut writer: impl Write,
+    validator: &Option<impl Validator>,
+) -> Result<String> {
+    let mut input = String::new();
+
+    loop {
+        match reader.read_line(&mut input)? {
+            0 => return Err(error::ReadlineError::Eof),
+            _ => {
+                input = apply_backspace_direct(&input);
+
+                match validator.as_ref() {
+                    None => break,
+                    Some(v) => {
+                        struct SimpleInvoke<'a> {
+                            s: &'a str,
+                        }
+
+                        impl<'a> keymap::Invoke for SimpleInvoke<'a> {
+                            fn input(&self) -> &str {
+                                self.s
+                            }
+                        }
+
+                        let mut ctx = SimpleInvoke {
+                            // Strip trailing newline for validation
+                            s: input
+                                .strip_suffix('\n')
+                                .map(|s| s.strip_suffix('\r').unwrap_or(s))
+                                .unwrap_or(&input),
```

Comment:
Are you sure ? `Validator` should support '\n' in input.

Reply:
The validator does support newlines. The issue here is that in interactive mode the validator is run without the trailing newline, so I just want this to be consistent.
I am guessing in interactive mode the newline is only added after the validator is run.
The newline here is not removed from the buffer, and it will end up in the validator input if the user is entering multiple lines.

Maybe the newline should be removed if the validator returns `Invalid` :thinking: 
In an earlier commit I was removing the trailing newline in the buffer all the time and adding it back on in the `Incomplete` case. The reason I changed it is that I was dropping `\r` while doing so.
`Invalid` is an awkward case to handle in non TTY mode anyway. :sweat_smile:

Reply:
Ok, I believe I have fixed that edge case. I think this issue is a matter of thinking of `\n` as a control character in the input sequence and not as a character that happens to be part of the input.